string sender
int32 px
int32 py
